stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+        // available payload bytes in a zero-filled array of MTU size. This makes the logic
+        // for filling the mailbox simpler as the payload pointer can be to variable-size
+        // data or even be a NULL pointer.
+        static uint8_t scratch_data[8] = {0};  // Not strictly needed to zero the storage.
+        if (payload_size > 0U)                 // The check is needed to avoid calling memcpy() with a NULL pointer.
+        {
+            // Clang-Tidy raises an error recommending the use of memcpy_s() instead.
+            // We ignore it because the safe functions are poorly supported; reliance on them may limit the portability.
+            (void) memcpy(&scratch_data[0], payload, payload_size);  // NOLINT
+        }
+
+        mb->TDTR = payload_size;  // DLC equals data length except in CAN FD
+
+        mb->TDHR = (((uint32_t) scratch_data[7]) << 24U) | (((uint32_t) scratch_data[6]) << 16U) |
+                   (((uint32_t) scratch_data[5]) << 8U) | (((uint32_t) scratch_data[4]) << 0U);
+        mb->TDLR = (((uint32_t) scratch_data[3]) << 24U) | (((uint32_t) scratch_data[2]) << 16U) |
+                   (((uint32_t) scratch_data[1]) << 8U) | (((uint32_t) scratch_data[0]) << 0U);
+
+        mb->TIR = convertFrameIDToRegister(extended_can_id) | BXCAN_TIR_TXRQ;  // Go.
+
+        // The frame is now enqueued and pending transmission.
+        out_status = true;
+    }
+
+    return out_status;
+}
+
+bool bxCANPop(const uint8_t   iface_index,          //
+              uint32_t* const out_extended_can_id,  //
+              size_t* const   out_payload_size,     //
+              void* const     out_payload)
+{
+    // Error management and reporting flags.
+    bool input_ok   = true;   // Initialize as no errors. Detected errors will set this to false.
+    bool out_status = false;  // Initialize as failure. It will be set true in case of success.
+
+    // Select the appropriate CAN interface base address and error flag.
+    volatile BxCANType* bxcan_base  = NULL;
+    bool*               error_iface = NULL;
+    if (iface_index == 0U)
+    {
+        bxcan_base  = BXCAN1;
+        error_iface = &g_error_iface0;
+    }
+    else if ((iface_index == 1U) && (BXCAN_MAX_IFACE_INDEX == 1U))
+    {
+        bxcan_base  = BXCAN2;
+        error_iface = &g_error_iface1;
+    }
+    else
+    {
+        input_ok = false;  // Invalid CAN interface number.
+    }
+
+    // Validate the output pointers.
+    if ((out_extended_can_id == NULL) || (out_payload_size == NULL) || (out_payload == NULL))
+    {
+        input_ok = false;  // Output pointers can not be NULL.
+    }
+
+    // Only try to receive data if the input validation was successful.
+    if (input_ok == true)
+    {
+        // This function must be polled periodically, so we use this opportunity to do it.
+        processErrorStatus(bxcan_base, error_iface);
+
+        static volatile uint32_t* RFxR[2];
```

Comment:
You really should lay off static things, global state it's not good for you.

```suggestion
        volatile uint32_t* const RFxR[2] = {&bxcan_base->RF0R, &bxcan_base->RF1R};
```

Reply:
There are various cases of "static const". Why are those fine? They don't always need to retain their value between calls.

Reply:
> Why are those fine?

Replied above.